File: machine/cpp/machine/scheduler/device.cpp
#include "machine/scheduler/device.h"

#include "machine/cpu/cpu.h"

#include <limits>

namespace bmsx {

DeviceScheduler::DeviceScheduler(CPU& cpu)
	: m_cpu(cpu) {
}

uint32_t DeviceScheduler::nextTimerGeneration(uint32_t value) {
	const uint32_t next = value + 1u;
	return next == 0u ? 1u : next;
}

void DeviceScheduler::setNowCycles(i64 nowCycles) {
	m_schedulerNowCycles = nowCycles;
}

void DeviceScheduler::reset() {
	clearTimerHeap();
	m_schedulerNowCycles = 0;
	m_schedulerSliceActive = false;
	m_activeSliceBaseCycle = 0;
	m_activeSliceBudgetCycles = 0;
	m_activeSliceTargetCycle = 0;
	m_deviceServiceTimerGeneration.fill(0);
}

i64 DeviceScheduler::currentNowCycles() const {
	if (!m_schedulerSliceActive) {
		return m_schedulerNowCycles;
	}
	return m_activeSliceBaseCycle + (m_activeSliceBudgetCycles - m_cpu.instructionBudgetRemaining);
}

void DeviceScheduler::beginCpuSlice(int sliceBudget) {
	m_schedulerSliceActive = true;
	m_activeSliceBaseCycle = m_schedulerNowCycles;
	m_activeSliceBudgetCycles = sliceBudget;
	m_activeSliceTargetCycle = m_schedulerNowCycles + sliceBudget;
}

void DeviceScheduler::endCpuSlice() {
	m_schedulerSliceActive = false;
}

void DeviceScheduler::advanceTo(i64 nowCycles) {
	m_schedulerNowCycles = nowCycles;
}

i64 DeviceScheduler::nextDeadline() {
	discardStaleTopTimers();
	if (m_timerDeadlines.empty()) {
		return std::numeric_limits<i64>::max();
	}
	return m_timerDeadlines[0];
}

bool DeviceScheduler::hasDueTimer() {
	discardStaleTopTimers();
	return !m_timerDeadlines.empty() && m_timerDeadlines[0] <= m_schedulerNowCycles;
}

uint8_t DeviceScheduler::popDueTimer() {
	const uint8_t deviceKind = m_timerDeviceKinds[0];
	removeTopTimer();
	return deviceKind;
}

void DeviceScheduler::scheduleDeviceService(uint8_t deviceKind, i64 deadlineCycles) {
	const uint32_t generation = nextTimerGeneration(m_deviceServiceTimerGeneration[deviceKind]);
	m_deviceServiceTimerGeneration[deviceKind] = generation;
	pushTimer(deadlineCycles, deviceKind, generation);
	requestYieldForEarlierDeadline(deadlineCycles);
}

void DeviceScheduler::cancelDeviceService(uint8_t deviceKind) {
	uint32_t& generation = m_deviceServiceTimerGeneration[deviceKind];
	generation = nextTimerGeneration(generation);
}

void DeviceScheduler::clearTimerHeap() {
	m_timerDeadlines.clear();
	m_timerDeviceKinds.clear();
	m_timerGenerations.clear();
}

void DeviceScheduler::pushTimer(i64 deadline, uint8_t deviceKind, uint32_t generation) {
	size_t index = m_timerDeadlines.size();
	m_timerDeadlines.push_back(deadline);
	m_timerDeviceKinds.push_back(deviceKind);
	m_timerGenerations.push_back(generation);
	while (index > 0) {
		const size_t parent = (index - 1u) >> 1u;
		if (m_timerDeadlines[parent] <= deadline) {
			break;
		}
		m_timerDeadlines[index] = m_timerDeadlines[parent];
		m_timerDeviceKinds[index] = m_timerDeviceKinds[parent];
		m_timerGenerations[index] = m_timerGenerations[parent];
		index = parent;
	}
	m_timerDeadlines[index] = deadline;
	m_timerDeviceKinds[index] = deviceKind;
	m_timerGenerations[index] = generation;
}

void DeviceScheduler::removeTopTimer() {
	const size_t lastIndex = m_timerDeadlines.size() - 1u;
	const i64 deadline = m_timerDeadlines[lastIndex];
	const uint8_t deviceKind = m_timerDeviceKinds[lastIndex];
	const uint32_t generation = m_timerGenerations[lastIndex];
	m_timerDeadlines.pop_back();
	m_timerDeviceKinds.pop_back();
	m_timerGenerations.pop_back();
	if (lastIndex == 0u) {
		return;
	}
	size_t index = 0u;
	while (index < (lastIndex >> 1u)) {
		size_t child = (index << 1u) + 1u;
		if (child + 1u < lastIndex && m_timerDeadlines[child + 1u] < m_timerDeadlines[child]) {
			child += 1u;
		}
		if (m_timerDeadlines[child] >= deadline) {
			break;
		}
		m_timerDeadlines[index] = m_timerDeadlines[child];
		m_timerDeviceKinds[index] = m_timerDeviceKinds[child];
		m_timerGenerations[index] = m_timerGenerations[child];
		index = child;
	}
	m_timerDeadlines[index] = deadline;
	m_timerDeviceKinds[index] = deviceKind;
	m_timerGenerations[index] = generation;
}

void DeviceScheduler::discardStaleTopTimers() {
	while (!m_timerDeadlines.empty()) {
		if (m_timerGenerations[0] == m_deviceServiceTimerGeneration[m_timerDeviceKinds[0]]) {
			return;
		}
		removeTopTimer();
	}
}

void DeviceScheduler::requestYieldForEarlierDeadline(i64 deadlineCycles) {
	if (!m_schedulerSliceActive) {
		return;
	}
	if (deadlineCycles > m_activeSliceTargetCycle) {
		return;
	}
	m_cpu.requestYield();
}

} // namespace bmsx

File: machine/ts/machine/scheduler/device.ts
import type { CPU, RunResult } from '../cpu/cpu';

export const DEVICE_SERVICE_GEO = 1;
export const DEVICE_SERVICE_DMA = 2;
export const DEVICE_SERVICE_APU = 3;
export const DEVICE_SERVICE_GPU = 4;
export const DEVICE_SERVICE_APU_TRANSFER = 5;
export const DEVICE_SERVICE_SYSTEM = 6;
export const DEVICE_SERVICE_GTE = 7;
export const DEVICE_SERVICE_IMGDEC = 8;

const DEVICE_SERVICE_KIND_COUNT = DEVICE_SERVICE_IMGDEC + 1;

function nextTimerGeneration(value: number): number {
	const next = (value + 1) >>> 0;
	return next === 0 ? 1 : next;
}

export class DeviceScheduler {
	private schedulerNowCycles = 0;
	private schedulerSliceActive = false;
	private activeSliceBaseCycle = 0;
	private activeSliceBudgetCycles = 0;
	private activeSliceTargetCycle = 0;
	private readonly timerDeadlines: number[] = [];
	private readonly timerDeviceKinds: number[] = [];
	private readonly timerGenerations: number[] = [];
	private readonly deviceServiceTimerGeneration = new Uint32Array(DEVICE_SERVICE_KIND_COUNT);

	public constructor(private readonly cpu: CPU) {
	}

	public get nowCycles(): number {
		return this.schedulerNowCycles;
	}

	public setNowCycles(nowCycles: number): void {
		this.schedulerNowCycles = nowCycles;
	}

	public reset(): void {
		this.clearTimerHeap();
		this.schedulerNowCycles = 0;
		this.schedulerSliceActive = false;
		this.activeSliceBaseCycle = 0;
		this.activeSliceBudgetCycles = 0;
		this.activeSliceTargetCycle = 0;
		this.deviceServiceTimerGeneration.fill(0);
	}

	public currentNowCycles(): number {
		if (!this.schedulerSliceActive) {
			return this.schedulerNowCycles;
		}
		return this.activeSliceBaseCycle + (this.activeSliceBudgetCycles - this.cpu.instructionBudgetRemaining);
	}

	public isCpuSliceActive(): boolean {
		return this.schedulerSliceActive;
	}

	public beginCpuSlice(sliceBudget: number): void {
		this.schedulerSliceActive = true;
		this.activeSliceBaseCycle = this.schedulerNowCycles;
		this.activeSliceBudgetCycles = sliceBudget;
		this.activeSliceTargetCycle = this.schedulerNowCycles + sliceBudget;
	}

	public endCpuSlice(): void {
		this.schedulerSliceActive = false;
	}

	public runCpuSlice(targetDepth: number, sliceBudget: number): RunResult {
		this.beginCpuSlice(sliceBudget);
		try {
			return this.cpu.runUntilDepth(targetDepth, sliceBudget);
		} finally {
			this.endCpuSlice();
		}
	}

	public advanceTo(nowCycles: number): void {
		this.schedulerNowCycles = nowCycles;
	}

	public nextDeadline(): number {
		this.discardStaleTopTimers();
		if (this.timerDeadlines.length === 0) {
			return Number.MAX_SAFE_INTEGER;
		}
		return this.timerDeadlines[0]!;
	}

	public hasDueTimer(): boolean {
		this.discardStaleTopTimers();
		return this.timerDeadlines.length > 0 && this.timerDeadlines[0]! <= this.schedulerNowCycles;
	}

	public popDueTimer(): number {
		const deviceKind = this.timerDeviceKinds[0]!;
		this.removeTopTimer();
		return deviceKind;
	}

	public scheduleDeviceService(deviceKind: number, deadlineCycles: number): void {
		const generation = nextTimerGeneration(this.deviceServiceTimerGeneration[deviceKind]!);
		this.deviceServiceTimerGeneration[deviceKind] = generation;
		this.pushTimer(deadlineCycles, deviceKind, generation);
		this.requestYieldForEarlierDeadline(deadlineCycles);
	}

	public cancelDeviceService(deviceKind: number): void {
		this.deviceServiceTimerGeneration[deviceKind] = nextTimerGeneration(this.deviceServiceTimerGeneration[deviceKind]!);
	}

	private clearTimerHeap(): void {
		this.timerDeadlines.length = 0;
		this.timerDeviceKinds.length = 0;
		this.timerGenerations.length = 0;
	}

	// start repeated-sequence-acceptable -- Scheduler heap moves three parallel timer columns inline; helper calls would sit on the timer hot path.
	private pushTimer(deadline: number, deviceKind: number, generation: number): void {
		let index = this.timerDeadlines.length;
		this.timerDeadlines[index] = deadline;
		this.timerDeviceKinds[index] = deviceKind;
		this.timerGenerations[index] = generation;
		while (index > 0) {
			const parent = (index - 1) >> 1;
			if (this.timerDeadlines[parent]! <= deadline) {
				break;
			}
			this.timerDeadlines[index] = this.timerDeadlines[parent]!;
			this.timerDeviceKinds[index] = this.timerDeviceKinds[parent]!;
			this.timerGenerations[index] = this.timerGenerations[parent]!;
			index = parent;
		}
		this.timerDeadlines[index] = deadline;
		this.timerDeviceKinds[index] = deviceKind;
		this.timerGenerations[index] = generation;
	}

	private removeTopTimer(): void {
		const lastIndex = this.timerDeadlines.length - 1;
		const deadline = this.timerDeadlines[lastIndex]!;
		const deviceKind = this.timerDeviceKinds[lastIndex]!;
		const generation = this.timerGenerations[lastIndex]!;
		this.timerDeadlines.length = lastIndex;
		this.timerDeviceKinds.length = lastIndex;
		this.timerGenerations.length = lastIndex;
		if (lastIndex === 0) {
			return;
		}
		let index = 0;
		while (index < (lastIndex >> 1)) {
			let child = (index << 1) + 1;
			if (child + 1 < lastIndex && this.timerDeadlines[child + 1]! < this.timerDeadlines[child]!) {
				child += 1;
			}
			if (this.timerDeadlines[child]! >= deadline) {
				break;
			}
			this.timerDeadlines[index] = this.timerDeadlines[child]!;
			this.timerDeviceKinds[index] = this.timerDeviceKinds[child]!;
			this.timerGenerations[index] = this.timerGenerations[child]!;
			index = child;
		}
		this.timerDeadlines[index] = deadline;
		this.timerDeviceKinds[index] = deviceKind;
		this.timerGenerations[index] = generation;
	}
	// end repeated-sequence-acceptable

	private discardStaleTopTimers(): void {
		while (this.timerDeadlines.length > 0) {
			if (this.timerGenerations[0] === this.deviceServiceTimerGeneration[this.timerDeviceKinds[0]!]) {
				return;
			}
			this.removeTopTimer();
		}
	}

	private requestYieldForEarlierDeadline(deadlineCycles: number): void {
		if (!this.schedulerSliceActive) {
			return;
		}
		if (deadlineCycles > this.activeSliceTargetCycle) {
			return;
		}
		this.cpu.requestYield();
	}
}
